sstables/kl/reader_impl.hh

```diff
@@ -170,277 +144,248 @@ private:
+    semaphore_type _production_sem;
+    future<> _production;
+    bool* _production_end;
 public:
     using consumer = row_consumer;
-    bool non_consuming() const {
-        return (((_state == state::DELETION_TIME_3)
-                || (_state == state::CELL_VALUE_BYTES_2)
-                || (_state == state::ATOM_START_2)
-                || (_state == state::ATOM_MASK_2)
-                || (_state == state::STOP_THEN_ATOM_START)
-                || (_state == state::COUNTER_CELL_2)
-                || (_state == state::RANGE_TOMBSTONE_4)
-                || (_state == state::EXPIRING_CELL_3)));
-    }
 
-    // process() feeds the given data into the state machine.
-    // The consumer may request at any point (e.g., after reading a whole
-    // row) to stop the processing, in which case we trim the buffer to
-    // leave only the unprocessed part. The caller must handle calling
-    // process() again, and/or refilling the buffer, as needed.
-    data_consumer::processing_result process_state(temporary_buffer<char>& data) {
-        try {
-            return do_process_state(data);
-        } catch (malformed_sstable_exception& exp) {
-            throw malformed_sstable_exception(exp.what(), _sst->get_filename());
+private:
+    future<> produce_64() {
+        co_await read_nonempty();
+        data_consumer::read_status rs = read_64(_data);
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
         }
+        co_return;
     }
-private:
-    data_consumer::processing_result do_process_state(temporary_buffer<char>& data) {
-#if 0
-        // Testing hack: call process() for tiny chunks separately, to verify
-        // that primitive types crossing input buffer are handled correctly.
-        constexpr size_t tiny_chunk = 1; // try various tiny sizes
-        if (data.size() > tiny_chunk) {
-            for (unsigned i = 0; i < data.size(); i += tiny_chunk) {
-                auto chunk_size = std::min(tiny_chunk, data.size() - i);
-                auto chunk = data.share(i, chunk_size);
-                if (process(chunk) == row_consumer::proceed::no) {
-                    data.trim_front(i + chunk_size - chunk.size());
-                    return row_consumer::proceed::no;
-                }
-            }
-            data.trim(0);
-            return row_consumer::proceed::yes;
+    future<> produce_32() {
+        co_await read_nonempty();
+        data_consumer::read_status rs = read_32(_data);
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
         }
-#endif
-        sstlog.trace("data_consume_row_context {}: state={}, size={}", fmt::ptr(this), static_cast<int>(_state), data.size());
```

Comment:
Please keep the trace logging in some way.

Reply:
I think I deleted this by mistake when clearing up my temporary debug, will remember to keep it in the final version